[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46738 [quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24486972](https://our.internmc.facebook.com/intern/diff/D24486972)

jerryzh168 committed:
[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24486972](https://our.internmc.facebook.com/intern/diff/D24486972)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24486972](https://our.internmc.facebook.com/intern/diff/D24486972)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1458,10 +1499,17 @@ def forward(self, x):
-        qconfig_dict = {'': default_qconfig}
-        prepared = prepare_fx(m, qconfig_dict)
+        qconfig_dict = {'': default_qat_qconfig}
+        prepared = prepare_qat_fx(m, qconfig_dict)
```
What happens for the case of PTQ? We should not be inserting observers in that case. Is that being tested?

jerryzh168 replied:
we'll insert fixedqparamfakequant in that case as well

jerryzh168 replied:
I can add a test for ptq

raghuramank100 commented on torch/quantization/fx/pattern_utils.py:
```diff
@@ -14,17 +14,30 @@ def get_default_fusion_patterns():
+def register_quant_pattern(pattern, output_activation_post_process=None):
     def insert(fn):
         DEFAULT_QUANTIZATION_PATTERNS[pattern] = fn
+        if output_activation_post_process is not None:
```
When would output_activation_post_process be None?

jerryzh168 replied:
this is only used for specifying special activaiton_post_process (e.g. fixed qparam fake quant) for a pattern, it is None for all other cases.

raghuramank100 commented on torch/quantization/fx/quantize.py:
```diff
@@ -346,7 +345,7 @@ def _prepare(self, model, qconfig_dict, inplace, prepare_custom_config_dict, is_
         # find _inputs_ to matched nodes that are not quantized, these
         # have to be quantized, which requires measuring stats,
-        # initialize an DefaultQuant object for each
+        # initialize an DefaultQuantizeHandler object for each
```
The function _prepare is really long, can you refactor it into smaller functions in a later PR?

jerryzh168 replied:
sure

raghuramank100 commented on test/quantization/test_quantize.py:
```diff
@@ -1247,6 +1248,36 @@ def forward(self, x):
+                self.dequant = DeQuantStub()
+
+            def forward(self, x):
+                x = self.quant(x)
```
Does the approach here also work for cases where the scale and zero-point are not fixed? i.e for a hardTanh, the scale and zero-point depend on the arguments to init.

jerryzh168 replied:
maybe, this allows to provide special fake quantize(activation_post_process) for a specific pattern (module/functional/torch op), could you write down the details, is it like following?
```
fake_quant = HardTanhFakeQuantize(hardtanh_instance.params)
```
I think this should be too hard to support in current implementation.

raghuramank100 replied:
Actually for hardtanh, our implementation calls qclamp: https://github.com/pytorch/pytorch/blob/87a4baf616565aecd21884754772a0f03babfe2d/aten/src/ATen/native/quantized/cpu/qclamp.cpp#L24, which doesn't need LUTs. In this case, we should treat it like ReLU, i.e there is no fake-quant or observer needed at the output as there is no quantization that occurs inside the op.

jerryzh168 replied:
yeah this is how it is implemented right now: https://github.com/pytorch/pytorch/blob/master/torch/quantization/fx/quantization_patterns.py#L512

supriyar commented on torch/quantization/quantize.py:
```diff
@@ -9,14 +9,23 @@
+def is_activation_post_process(module):
+    return (isinstance(module, torch.quantization.ObserverBase) or
+            isinstance(module, torch.quantization.FakeQuantize) or
+            isinstance(module, torch.quantization.FakeQuantizeBase))
```
Do we plan to make FakeQuantize a subclass of FakeQuantizeBase?

jerryzh168 replied:
yes, will do that sometime in the future, maybe before end of this year.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24486972](https://our.internmc.facebook.com/intern/diff/D24486972)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24486972](https://our.internmc.facebook.com/intern/diff/D24486972)

[ghstack-poisoned]

supriyar commented on torch/quantization/quantize.py:
```diff
@@ -107,31 +116,34 @@ def add_observer_(module, qconfig_propagation_list=None, non_leaf_module_list=No
         if type(child) == nnq.FloatFunctional or type(child) == nnq.QFunctional:
-            if hasattr(child, 'qconfig') and child.qconfig is not None:
+            if needs_observation(child):
                 child.activation_post_process = get_activation_post_process(child.qconfig, device)
```
I think this callsite needs to be updated.

jerryzh168 replied:
why? this is handling FloatFunctional and we don't have special activation post process for it.

jerryzh168 replied:
oh, I guess I forgot about setting default argument for special_act_post_process for get_activation_post_process.

supriyar commented on test/quantization/test_quantize_fx.py:
```diff
@@ -1488,6 +1471,82 @@ def forward(self, x):
+            prepared = prepare(m, qconfig_dict)
+            # check the correct number of activation_post_process is inserted
+            count_check = {
+                ns.call_module(FixedQParamsFakeQuantize) : 13,
```
Do we want to use the `FixedQParamsFakeQuantize` for PTQ as well? FakeQuantize is something we typically use during QAT.

supriyar replied:
Is there a way to fix scale and zero-point for PTQ without using this module? Or should we have a similar `FixedQParamsObserver`?

jerryzh168 replied:
yeah, actually, after second thought, I feel we don't need to insert anything for PTQ, since the information is not needed and also there is no point to have fake quantize module in the observed model.

jerryzh168 replied:
Thanks Supriya, updated the PR.

supriyar replied:
Looks good. Could you clarify the flow for PTQ? How does the scale/zero_point for the fixed qparams ops get calculated?

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24486972](https://our.internmc.facebook.com/intern/diff/D24486972)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Support sigmoid/hardsigmoid/tanh in qat"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24486972](https://our.internmc.facebook.com/intern/diff/D24486972)

[ghstack-poisoned]
